File: backend/app/routers/public.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy import func
from sqlalchemy.orm import Session

from app import schemas
from app.database import get_db
from app.models import (
    CounsellorProfile,
    CounsellorProfileStatus as CounsellorProfileStatusModel,
    Resource,
    ResourceStatus,
    SessionRequest,
    SessionRequestStatus as SessionRequestStatusModel,
)

router = APIRouter(prefix="/public", tags=["Public"])


@router.get("/counsellors/featured", response_model=schemas.FeaturedCounsellorsResponse)
def get_featured_counsellors(
    limit: int = Query(default=12, ge=1, le=50),
    db: Session = Depends(get_db),
):
    """Unauthenticated landing carousel (audit §5.1).

    Returns active counsellor profiles. Only users that actually have an active
    profile are included; an empty result is a valid empty list, never an error.
    """
    profiles = (
        db.query(CounsellorProfile)
        .filter(CounsellorProfile.status == CounsellorProfileStatusModel.active)
        .order_by(CounsellorProfile.id)
        .limit(limit)
        .all()
    )

    cards = [
        schemas.CounsellorCard(
            id=profile.user_id,
            short_name=profile.short_name,
            initials=profile.initials,
            photo_url=profile.photo_url,
            specialties=profile.specialties,
            bio=profile.bio,
            availability_note=profile.availability_note,
        )
        for profile in profiles
    ]
    return schemas.FeaturedCounsellorsResponse(counsellors=cards)


@router.get("/stats", response_model=schemas.PublicStatsResponse)
def get_public_stats(db: Session = Depends(get_db)):
    counsellor_count = (
        db.query(CounsellorProfile)
        .filter(CounsellorProfile.status == CounsellorProfileStatusModel.active)
        .count()
    )

    students_supported = (
        db.query(func.count(func.distinct(SessionRequest.student_id)))
        .filter(
            SessionRequest.status.in_(
                [
                    SessionRequestStatusModel.accepted,
                    SessionRequestStatusModel.completed,
                ]
            )
        )
        .scalar()
        or 0
    )

    sessions_completed = (
        db.query(SessionRequest)
        .filter(SessionRequest.status == SessionRequestStatusModel.completed)
        .count()
    )

    resources_published = (
        db.query(Resource)
        .filter(Resource.status == ResourceStatus.published)
        .count()
    )

    return schemas.PublicStatsResponse(
        counsellor_count=counsellor_count,
        students_supported=students_supported,
        sessions_completed=sessions_completed,
        resources_published=resources_published,
    )

File: frontend/src/api/public.js
import { apiFetch } from "./client";

export async function getFeaturedCounsellors({ status = "available", limit = 12 } = {}) {
  const params = new URLSearchParams();
  if (status) params.set("status", status);
  if (limit) params.set("limit", String(limit));
  const query = params.toString();
  const data = await apiFetch(`/public/counsellors/featured${query ? `?${query}` : ""}`, {
    auth: false,
  });
  return (data.counsellors ?? []).map(mapCounsellorCard);
}

export function mapCounsellorCard(item) {
  return {
    id: item.id,
    shortName: item.short_name,
    initials: item.initials,
    photoUrl: item.photo_url,
    specialties: item.specialties ?? [],
    bio: item.bio ?? "",
    availabilityNote: item.availability_note ?? "",
  };
}

export async function getPublicStats() {
  return apiFetch("/public/stats", { auth: false });
}

export function mapPublicStats(api) {
  return {
    counsellorCount: api.counsellor_count ?? 0,
    studentsSupported: api.students_supported ?? 0,
    sessionsCompleted: api.sessions_completed ?? 0,
    resourcesPublished: api.resources_published ?? 0,
  };
}

function formatStatCount(value) {
  if (value >= 100) return `${value}+`;
  return String(value);
}

export function formatPublicStatDisplay(stats) {
  return {
    counsellors: formatStatCount(stats.counsellorCount),
    studentsSupported: formatStatCount(stats.studentsSupported),
  };
}
